Clean/Circuit/StructuralLemmas.lean

```diff
@@ -0,0 +1,255 @@
+    (elaborated : ElaboratedCircuit F Input Output)
+    (Assumptions : Input F → Prop)
+    (Spec : Input F → Output F → Prop)
+    (subcircuit : FormalCircuit F Mid Output)
+    (prepare : Var Input F → Circuit F (Var Mid F))
+    (h_main_eq : ∀ input, elaborated.main input = prepare input >>= fun mid => subcircuit mid)
+    (h_output_eq : ∀ input offset, elaborated.output input offset =
+      subcircuit.output ((prepare input).output offset) (offset + (prepare input).localLength offset))
+    (h_spec_implication : ∀ input mid output,
+      Assumptions input →
+      subcircuit.Assumptions mid →
+      subcircuit.Spec mid output →
+      Spec input output)
+    (h_prepare_spec : ∀ offset env input_var input,
+      eval env input_var = input →
+      Assumptions input →
+      ConstraintsHold.Soundness env (prepare input_var |>.operations offset) →
+      let mid := eval env (prepare input_var |>.output offset)
+      subcircuit.Assumptions mid) :
+    Soundness F elaborated Assumptions Spec := by
+  intro offset env input_var input h_eval h_assumptions h_holds
+
+  -- Rewrite using the main equation
+  rw [h_main_eq] at h_holds
+  simp only [Circuit.bind_def, circuit_norm] at h_holds
+
+  -- Extract constraints for prepare and subcircuit
+  obtain ⟨h_holds_prepare, h_holds_subcircuit⟩ := h_holds
+
+  -- Get the intermediate values
+  let mid_var := (prepare input_var).output offset
+  let mid := eval env mid_var
+  let prepare_len := (prepare input_var).localLength offset
+
+  -- Prove the subcircuit assumptions hold
+  have h_mid_assumptions : subcircuit.Assumptions mid := by
+    apply h_prepare_spec offset env input_var input h_eval h_assumptions h_holds_prepare
+
+  -- The subcircuit soundness gives us the spec
+  simp only [subcircuit_norm, FormalCircuit.toSubcircuit] at h_holds_subcircuit
+
+  -- Apply subcircuit soundness
+  have h_subcircuit_spec : subcircuit.Spec mid (eval env (subcircuit.output mid_var (offset + prepare_len))) := by
+    exact h_holds_subcircuit h_mid_assumptions
+
+  -- Use the output equation
+  rw [h_output_eq]
+
+  -- Apply the spec implication
+  apply h_spec_implication input mid _ h_assumptions h_mid_assumptions h_subcircuit_spec
+
+/--
+Structural soundness lemma for composing two FormalCircuits.
+
+This lemma allows proving soundness of a composite circuit that consists of
+two FormalCircuits executed in sequence, where:
+1. The first circuit transforms Input to Mid
+2. The second circuit transforms Mid to Output
+3. The spec of the composite relates Input to Output
+
+This is useful for circuits that naturally decompose into two sequential stages.
+-/
+theorem soundness_compose_circuits
+    (elaborated : ElaboratedCircuit F Input Output)
+    (Assumptions : Input F → Prop)
+    (Spec : Input F → Output F → Prop)
+    (circuit1 : FormalCircuit F Input Mid)
+    (circuit2 : FormalCircuit F Mid Output)
+    (h_main_eq : ∀ input, elaborated.main input = circuit1 input >>= circuit2)
```

Comment:
Consider dropping `elaborated` and `h_main_eq` and stating the rest of the lemma about the `>>=` term.

Reply:
This has been done.